Node.js CLI tutorial (#363)
mongodb/docs-realm: source/tutorial/nodejs-cli.txt

tracker command line interface (CLI) that allows users to:

- Register themselves with email and password.
- Sign in to their account with their email and password.
- View, create, modify, and delete tasks.
- Watch for changes to the Task collection and get notified in the terminal 
  window when a change occurs.

This tutorial should take around 30 minutes to complete.

.. admonition:: Check Out the Quick Start
   :class: note
   
   If you prefer to explore on your own rather than follow a guided tutorial,
   check out the :ref:`Quick Start <node-client-quick-start>`. It
   includes copyable code examples and the essential information that you need
   to set up a MongoDB Realm application.

.. admonition:: Download the Complete Source Code
   :class: note
   
   We host this tutorial application's :github:`complete and ready-to-compile
   source code <mongodb-university/realm-tutorial/tree/main/node-cli>` on
   GitHub. Just follow the instructions in ``README.md`` to get started. Don't
   forget to update the ``config.js`` file with your App ID, which you can
   find in the {+ui+}.

Prerequisites
-------------

Before you begin, ensure you have:

> - `Node.js <https://nodejs.org/en/download/>`_ installed.

Author: Weird! It renders correctly for me. @cbush can you confirm?

> - `Node.js <https://nodejs.org/en/download/>`_ installed.

Reviewer: Looks good now: https://docs-mongodbcom-staging.corp.mongodb.com/realm/caleb/node_cli/tutorial/nodejs-cli.html#prerequisites

     - Displays the main menu of choices.
   
   * - output.js
     - Responsible for displaying text in the terminal window.
   
   * - tasks.js
     - Handles all task-related communication between the CLI and Realm. The 
       methods for listing, creating, editing, and deleting tasks live here.
   
   * - schemas.js
     - Contains the schema definitions for the collections used in this project.

   * - users.js
     - Handles Realm user authentication, including logging in, registering a 
       new user, and logging out.

   * - watch.js
     - Adds a listener to a Realm collection to notify the user when a change 
       occurs.

C. Connect to Your MongoDB Realm App
------------------------------------
To get the app working with your backend, you first need to add your Realm App ID 
to the ``config.js`` file. The ``config.js`` module exports a single property, 
``realmAppId``, which is currently set to "TODO". Change this value to your 
Realm App ID. 

.. note::

   To learn how to find your MongoDB Realm ``appId``, read the :doc:`Find
   Your App Id </get-started/find-your-app-id>` doc.

Once you have made that change, you now need to complete the code needed to 

> open a Realm. In ``index.js``, find the ``openRealm`` function. It will look 

Reviewer 2: ```suggestion
open a {+realm+}. In ``index.js``, find the ``openRealm`` function. It will look 
```

like this:

.. code-block:: js

   async function openRealm() {
      const config = {
         schema: [schemas.TaskSchema, schemas.UserSchema, schemas.ProjectSchema],
         sync: {
            user: users.getAuthedUser(),
            partitionValue: "myPartition",
         },
      };

>       // TODO: open a Realm with these configuration settings.

Reviewer 2: Lowercase refers to realm instance
```suggestion
      // TODO: open a realm with these configuration settings.
```

   }

> Replace the ``TODO`` line with a line of code that opens a Realm and assigns 

Reviewer 2: ```suggestion
Replace the ``TODO`` line with a line of code that opens a {+realm+} and assigns 
```

it to the ``realm`` property:

.. code-block:: js

   realm = await Realm.open(config);

At this point, your app is pointing to your backend and opens a connection 
to it when you start the app. However, users cannot yet log in, so let's update 
that code next.

D. Enable authentication
------------------------
In the ``users.js`` file, we have a ``logIn`` function that prompts the user for 
an email address and password, and then, within a try-catch block, creates an 
:js-sdk:`emailPassword <Realm.App.html#Realm.Credentials.html#.emailPassword>` 
credential and passes it to the Realm 
:js-sdk:`logIn() <Realm.App.html#logIn>` method. 

> Find the the following comment:

Reviewer 2: ```suggestion
Find the following comment:
```

.. code-block:: js

    // TODO: create new emailPassword credentials and call app.logIn(...)

And add the following code to create a ``emailPassword`` credential:

.. code-block:: js

    const credentials = Realm.Credentials.emailPassword(
      input.email,
      input.password
    );

Immediately below this, add the following code to call the 
``logIn()`` method:

.. code-block:: js

    const user = await app.logIn(credentials);

E. Implement the CRUD methods
-----------------------------
In the ``tasks.js`` file, there are a number of functions to handle typical  
CRUD functionality: ``getTasks``, ``getTask``, ``createTask``, ``deleteTask``, 
``editTask``, and ``changeStatus``, which is essentially a specific case of 

> ``editTask``. Each of these functions (except ``getTasks``) prompts the user, 

Reviewer 2: No comma because the subject didn't change in the next clause
```suggestion
``editTask``. Each of these functions (except ``getTasks``) prompts the user
```

Reviewer: Tiny thought: prompts the user for what? Guessing each function expects different input parameters. Maybe something like "Each ... prompts the user for input / a response / a necessary parameter", etc. Very minor.

and then makes the appropriate call to Realm. Your task is to implement the 

> calls to Realm. The following list shows each function and provides 

Reviewer 2: Replacement here would be:
```suggestion
calls to {+service-short+}. The following list shows each function and provides 
```

> guidance on how to complete the task.

Reviewer: Slight wordsmithing:

"The following list provides guidance on how to complete this task for each function:"

* ``getTasks``

>   To get all objects, call the :js-sdk:`obects() <Realm.html#objects>` method 

Reviewer 2: ```suggestion
  To get all objects, call the :js-sdk:`objects() <Realm.html#objects>` method 
```

  and pass in the name of the collection:
  
  .. code-block:: js
     

>      const tasks = await realm.objects("Task");

Reviewer 2: I don't think this should be awaited (may need to update the tutorial source)

Author: I tested without the `await` and it failed.

Reviewer 2: Failed how? This is not an async function. We need to dig into this, because something is wrong here.

* ``getTask``
  
  In the Tasks collection, a task's ``id`` field is the primary key, so 
  we call the :js-sdk:`objectForPrimaryKey() <Realm.html#objectForPrimaryKey>`  
  function to get a task by its Id.

  .. code-block:: js
     

>      let result = await realm.objectForPrimaryKey(

Reviewer 2: It's misleading to `await` this non-async function call

       "Task",
       new bson.ObjectID(task.id)
     );

* ``createTask``

  Whenever we modify an object in realm, we must do so within a 
  :ref:`transaction <node-open-a-transaction>`. 
  While you can do this by calling the 
  :js-sdk:`beginTransaction() <Realm.html#beginTransaction>` and 
  :js-sdk:`commitTransaction() <Realm.html#commitTransaction>`

>   methods, the

Reviewer 2: Is it worth mentioning begin/commit? Should we prefer to lead with the recommended approach (write)?

Reviewer: Hmm, good point here. I do think it could be OK to include this context: it's useful to know that, maybe, if I wanted to do similar elsewhere, but couldn't use `write()` for some reason, I now know I have options. Also, maybe users already know about these two transaction functions, but not that the `write()` function includes them. But, if so, I do think we need to slightly-more-strongly encourage use of `write()`. Maybe something like:

"While you could perform this task using the X and Y functions, the Z function handles these tasks automatically for us as part of its operation, so we'll use that here.

Of course, a perfectly valid alternative is not mentioning it at all, as per Bush's suggestion.

Author: I think it's still worth mentioning them, as its a valid approach, but I have demoted them to a note after the code.

Reviewer 2: It's kind of a legacy approach with its own problems. Not sure if we want to recommend it at all.

  :js-sdk:`write() <Realm.html#write>` method takes care of 
  transaction handling for us. Within the ``write`` function, we then call
  the :js-sdk:`create() <Realm.html#create>` function, passing in all of the 
  required properties:

  .. code-block:: js

     realm.write(() => {
        result = realm.create("Task", {
          _id: new bson.ObjectID(),
          _partition: "myPartition",
          name: task.name,
          status: task.status,
        });
     });

* ``deleteTask``
    

>   As with creating an object, you delete an object within a transaction. Again, 

Reviewer: Previous point reference modifying an object, let's reuse that here for consistency (since you use "as with" to link back). Also struggled with "As with creating ... you delete". How about:

"Deleting objects must also take place within a transaction. As with modifying an object, we'll use the `write()` function to handle the transaction for us, and then call the the `delete()` function within it:"

Or similar. Still not quite happy with that rearrangement.

  .. code-block:: js

     realm.write(() => {
       realm.delete(task);
       output.result("Task deleted.");
     });

* ``modifyTask``
    
  This function is called by both the ``editTask`` and ``changeStatus`` functions. 
  Like the ``createTask`` and ``deleteTask`` methods, when you change an object, 
  you do so within a transaction. Other than that, though, there is no specific 
  call to a Realm API to change an object. Rather, you change the *local* object 
  and Sync ensures the object is updated on the server.
  
  .. code-block:: js
    
      realm.write(() => {
        task = realm.objectForPrimaryKey("Task", new bson.ObjectID(answers.id));
        task[answers.key] = answers.value;
      });

  .. note::
    
     To learn more about Realm Sync, see :doc:`Sync Overview </sync/overview>`.

F. Run and Test
---------------
Once you have completed the code, you should run the app and check functionality. 

1. Open a terminal window and change to your app's directory.

#. Run the following commands to install all of the dependencies and start the 
   app:

   .. code-block:: sh
      
      ``npm install``

>       ``node index.js``

Reviewer: No need for double-backticks while within `code-block`s

#. Your terminal window will clear and you will see the initial menu prompting 
   you to log in or register as a new user:

   .. image:: /images/node-cli-start-screen.png
      :alt:  Initial menu
    
#. If you do not yet have a user account, enter an email and password, and the 
   system will create a new account and log you in. At this point, you should see 
   the main "menu" of choices. All of the options should now work for you except 
   the "watch" functionality, which we'll enable in the next section.

G. Implement the Collection Listener
------------------------------------
A very handy feature in Realm is the ability to add a listener to a collection. 
The listener notifies your app when a change occurs in the collection. For our 
CLI, we want to listen for changes on the ``Tasks`` collection, and when one 
occurs, display a notification in the console window. To do this:

1. In your text editor, open the ``watch.js`` file. 

#. Our internal ``listener`` function is a callback function that takes two 

>    parameters: the collection of tasks and a ``changes`` dictionary that is sent 

Reviewer 2: I would probably use "object" instead of "dictionary" in JS land.

Author: The SDK uses "dictionary", but I agree for most js readers "object" is clearer.

   from Realm. The ``changes`` dictionary has the following keys:

   * **deletions** -- returns the index of the deleted item before the deletion.

   * **insertions** -- returns the index of the new object.

>    * **modifications** -- returns the index of the modified object.

Reviewer: Each of these items is plural (insinuates containing multiple items), but described as containing only one item. Perhaps "returns the index of any deleted / modified objects / items"?

Probably minor.

Author: Yeah, that's an annoying part of the SDK. That's what they're called, but they fire with each individual change/addition/deletion.

#. In the ``watchForChanges`` function, you will get all of the objects in 
   the **Tasks** collection (just as we did in the ``getTasks`` function),  
   and then call the :js-sdk:`addListener() <Realm.Collection.html#addListener>` 
   method. Under the ``TODO`` comment, add the following code:

.. code-block:: js
   

>    const tasks = await realm.objects("Task");

Reviewer 2: No need to await.

Author: apparently there is, based on my testing.

   tasks.addListener(listener);

With this code update, you can now re-run the app and watch for changes. 

1. In the main menu, choose "Watch for changes". You will see that the UI 
   stays in a "waiting" state. 

2. Open a second terminal window and run the app again. In this second window, 
   try creating, modifying, and deleting tasks. Each time you perform an action, 

>    the first terminal window will update with a notification. 

Reviewer: It might be nice to present one example create|modify|delete task that is copy and pasteable to showcase this behavior immediately, such that even a user who had not, at this point, completed the tutorial for those commands would be able to see the behavior of this watch feature in full.

Author: Since it's all done through the UI, there's no code that can demonstrate this.

Reviewer: This section says: "in the second (terminal) window, try creating ..."

Terminal window = UI?

Oh! It's the ASCII art menu selection process! Ah I see, sorry, I understand now.

.. note::
   
   If you do not want to open 2 terminal windows, you can modify the code in 
   ``main.js``, within the ``case Choices.WatchForChanges`` handler. The 

>    inline code comment explains the change you can make. 

Reviewer 2: I don't know. Why wouldn't someone want to open another terminal window?

Author: I guess the point is more "if you want to keep working and just have alerts pop up...", so I'll rewrite.

   
   git clone git@github.com:mongodb-university/realm-tutorial.git

.. important::
   
   The ``main`` branch is a finished version of the app as it should look *after*
   you complete this tutorial. To walk through this tutorial, please check 
   out the ``todo`` branch:

   .. code-block:: shell
    
      git checkout todo

> 

Reviewer: Nice, good call.

        });
     });

  .. note::
    
    The ``write`` function replaces the need to call the 
    :js-sdk:`beginTransaction() <Realm.html#beginTransaction>`,  
    :js-sdk:`commitTransaction() <Realm.html#commitTransaction>`, and 
    :js-sdk:`cancelTransaction() <Realm.html#cancelTransaction>`

>     methods.

Reviewer: I think this, as a follow-up -- informational content outside the actionable path -- works well.

   the main "menu" of choices. All of the options should now work for you except 
   the "watch" functionality, which we'll enable in the next section.

.. admonition:: Reminder
   
   If something isn't working for you, you can check out the ``main`` branch of 
   this repo to compare your code with our finished solution. 

> 

Reviewer: 👍
